rfd/0170-package-distribution-v3.md

```diff
@@ -0,0 +1,426 @@
+  tool, similar to what we're doing today with Aptly.
+* The tool has some bugs and missing features that are critical to our technical
+  requirements.
+* We would still need to manage everything outside of this tool itself (such as
+  mutex locking), which we've been bitten by pretty regularly in the past.
+* The tooling for this would be built around yet another language, complicating
+  the maintenance of our release process further.
+
+Ultimately we would _need_ to fork this tool and put in a bunch of work just to
+get it into a state where we could use it, and then we'd need to continue to
+maintain it long-term. I do not believe that doing this would save us
+significant effort over implementing the entire deb and rpm repo specs from
+scratch in Go.
+
+I estimate that we would need a minimum of a full engineering-quarter worth of
+work for either of these routes. Excluding the opportunity cost of not being
+able to work on other projects, I estimate that the development costs of this
+would be $100k worth of engineering at absolute minimum. The infra and
+maintenance costs would likely be similar to our current in-house solution,
+costing a minimum of $300k/year on top of the initial development costs.
+
+##### Pulp-based tooling
+
+[Pulp](https://pulpproject.org/) would allow us to focus less on the repo
+generation logic, and more on how the tool integrates with our business needs.
+It consists of several services that we would need to host (front end, API
+service, plugin service, database, cache, storage, CDN). It is actively
+maintained by Red Hat.
+
+However, it has several downsides:
+* It's written in Python, so any changes to the (collective) Pulp service would
+  likely also need to be written/maintained with Python.
+* It's plugin based, so each repo type (APT, YUM, etc.) has an entirely
+  different interface from the ground up. Despite being one collective service,
+  the requirements, capabilities, and APIs for each are wildly different. For
+  example, the YUM plugin supports a S3 backend while the APT plugin does not.
+* The documentation is severely lacking, and hard to use. There is no paid
+  support for this tool.
+
+It is my strong opinion that this would be a maintenance and security nightmare.
+
+#### Third party solutions
+
+There are many options for third party solutions (both SaaS and on-prem) to our
+problem:
+
+* AWS CodeArtifact
+* Azure Artifacts
+* Cloudsmith (requires providing signing key)
+* PackageCloud (requires providing signing key)
+* GCP Artifact Registry (requires providing signing key)
+* GitLab
+* Gitea
+* ProGet
+* Artifactory
+* Nexus Repository
+
+I've taken a look at all of them and demoed a few. Here's what I found.
+
+##### AWS CodeArtifact and Azure Artifacts
+
+Both AWS and Azure offer package hosting solutions. An AWS service would be
+great, as we already trust AWS with many of our sensitive secrets.
+Unfortunately, neither of these products support APT or YUM repos.
```

Comment:
It looks like they are continuously add support for more package types. But I couldn't find a roadmap for when/if they might add support for deb/rpm. If they did end up adding support for deb/rpm, would CodeArtifacts be recommended over Artifactory?

Would there be a way to build around the generic package type that CodeArtifacts supports?

Reply:
> Would there be a way to build around the generic package type that CodeArtifacts supports?

No not currently. CodeArtifact is basically a wrapper for S3 + cloudfront that also provides tools for package repo metadata. The primary problem we currently have is generating that metadata, so building this ourselves would gain us little.